rotkehlchen/chain/evm/decoding/curve_lend/decoder.py

```diff
@@ -0,0 +1,617 @@
+            context: DecoderContext,
+    ) -> tuple['EvmEvent | None', 'EvmEvent | None']:
+        """Decode events associated with a withdrawal.
+        Returns out_event and in_event in a tuple to be used for reordering."""
+        if (tokens_and_amounts := self._get_vault_event_tokens_and_amounts(
+                vault_address=context.tx_log.address,
+                context=context,
+        )) is None:
+            log.error(f'Failed to find tokens and amounts for Curve vault withdrawal transaction {context.transaction}')  # noqa: E501
+            return None, None
+
+        vault_token, underlying_token, shares_amount, assets_amount = tokens_and_amounts
+        out_event, in_event = None, None
+        for event in context.decoded_events:
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.address == ZERO_ADDRESS and
+                event.asset == vault_token and
+                event.balance.amount == shares_amount
+            ):
+                event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                event.notes = f'Return {shares_amount} {vault_token.symbol} to a Curve vault'
+                event.counterparty = CPT_CURVE
+                out_event = event
+            elif (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.asset == underlying_token and
+                event.balance.amount == assets_amount
+            ):
+                event.event_type = HistoryEventType.WITHDRAWAL
+                event.event_subtype = HistoryEventSubType.REMOVE_ASSET
+                event.notes = f'Withdraw {assets_amount} {underlying_token.symbol} from a Curve vault'  # noqa: E501
+                event.counterparty = CPT_CURVE
+                in_event = event
+
+            if out_event is not None and in_event is not None:
+                break
+
+        return out_event, in_event
+
+    def _decode_vault_events(self, context: DecoderContext) -> DecodingOutput:
+        """Decode events from Curve vaults."""
+        if context.tx_log.topics[0] == DEPOSIT_TOPIC:
+            out_event, in_event = self._decode_deposit(context=context)
+        elif context.tx_log.topics[0] == WITHDRAW_TOPIC:
+            out_event, in_event = self._decode_withdraw(context=context)
+        else:
+            return DEFAULT_DECODING_OUTPUT
+
+        if out_event is None or in_event is None:
+            log.error(f'Failed to find both out and in events for Curve vault transaction {context.transaction}')  # noqa: E501
+            return DEFAULT_DECODING_OUTPUT
+
+        maybe_reshuffle_events(
+            ordered_events=[out_event, in_event],
+            events_list=context.decoded_events,
+        )
+        return DEFAULT_DECODING_OUTPUT
+
+    def _get_possibly_cached_vault_address(
```

Comment:
Think this agrees more with the naming we use in the codebase, right? And means same thing?

```suggestion
    def _maybe_get_cached_vault_address(
```